pkg/tracing/tracer_otel.go

```diff
@@ -0,0 +1,93 @@
+package tracing
+
+import (
+	"context"
+
+	"github.com/justtrackio/gosoline/pkg/cfg"
+	"github.com/justtrackio/gosoline/pkg/log"
+	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+func init() {
+	AddTracerProvider(ProviderOtel, NewOtelTracer)
+}
+
+const (
+	instrumentationName    = "https://github.com/justtrackio/gosoline"
+	instrumentationVersion = "v0.8.0"
+)
+
+type otelTracer struct {
+	tracer trace.Tracer
+}
+
+func NewOtelTracer(ctx context.Context, config cfg.Config, logger log.Logger) (Tracer, error) {
+	traceProvider, err := ProvideOtelTraceProvider(ctx, config, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	tracer := traceProvider.Tracer(
+		instrumentationName,
+		trace.WithInstrumentationVersion(instrumentationVersion),
+		trace.WithSchemaURL(semconv.SchemaURL),
+	)
+
+	return NewOtelTracerWithInterfaces(tracer), nil
+}
+
+func NewOtelTracerWithInterfaces(tracer trace.Tracer) *otelTracer {
+	return &otelTracer{
+		tracer: tracer,
+	}
+}
+
+func (t *otelTracer) StartSubSpan(ctx context.Context, name string) (context.Context, Span) {
+	ctx, span := t.tracer.Start(ctx, name)
+
+	return newOtelSpan(ctx, span)
+}
+
+func (t *otelTracer) StartSpan(name string) (context.Context, Span) {
+	return t.StartSubSpan(context.Background(), name)
+}
+
+func (t *otelTracer) StartSpanFromContext(ctx context.Context, name string) (context.Context, Span) {
+	if parentSpan := GetSpanFromContext(ctx); parentSpan != nil {
+		parentTrace := parentSpan.GetTrace()
+
+		return t.spanFromTrace(ctx, parentTrace, name)
+	}
+
+	if ctxTrace := GetTraceFromContext(ctx); ctxTrace != nil {
+		return t.spanFromTrace(ctx, ctxTrace, name)
+	}
+
+	return t.StartSubSpan(ctx, name)
+}
+
+func (t *otelTracer) spanFromTrace(ctx context.Context, trc *Trace, name string) (context.Context, Span) {
+	var tFlags trace.TraceFlags
+	if trc.GetSampled() {
+		tFlags = trace.FlagsSampled
+	}
+
+	tID, err := trace.TraceIDFromHex(trc.GetTraceId())
+	if err != nil {
+		tID = trace.TraceID{}
```

Comment:
So if we get an error, we get a parially filled object we have to zero out? Is this a behavior we should have a comment about?

Reply:
added a comment and also a warning log if that happens, under normal circumstances this should never happen